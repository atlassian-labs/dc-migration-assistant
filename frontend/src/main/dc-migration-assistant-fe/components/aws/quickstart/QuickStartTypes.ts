/*
 * Copyright 2020 Atlassian
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export type QuickStartParameterYamlNode = {
    Type: string;
    Default?: string | number | boolean;
    Description: string;
    AllowedValues?: Array<string | boolean>;
    ConstraintDescription?: string;
    AllowedPattern?: string;
    MaxLength?: number;
    MinLength?: number;
    MaxValue?: number;
    MinValue?: number;
    NoEcho?: boolean;
};

export type QuickstartParamLabelYamlNode = {
    default: string;
};

export type QuickstartParamGroupYamlNode = {
    Label: QuickstartParamLabelYamlNode;
    Parameters: Array<string>;
};

export type QuickstartParameterGroup = {
    groupLabel: string;
    shouldExpandGroupOnLoad: boolean;
    parameters: Array<QuickstartParameter>;
};

export type QuickstartParameterProperties = QuickStartParameterYamlNode;

export type QuickstartParameter = {
    paramKey: string;
    paramLabel: string;
    paramProperties: QuickstartParameterProperties;
};
