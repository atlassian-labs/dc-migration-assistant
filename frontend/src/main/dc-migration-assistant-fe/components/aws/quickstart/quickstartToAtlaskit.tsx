/*
 * Copyright 2020 Atlassian
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement } from 'react';
import Select, { AsyncSelect, OptionType } from '@atlaskit/select';
import Toggle from '@atlaskit/toggle';
import TextField from '@atlaskit/textfield';
import { ErrorMessage, Field, HelperMessage } from '@atlaskit/form';
import { I18n } from '@atlassian/wrm-react-i18n';

// eslint-disable-next-line import/extensions
import { QuickstartParameter } from './QuickStartTypes';

type FormElementGenerator = (
    defaultProps: Record<string, string>,
    param: QuickstartParameter
) => ReactElement;
type InputProps = Record<string, boolean | number | string | Function>;

const createAZSelection: FormElementGenerator = (defaultFieldProps, param) => {
    // TODO: This should be queried via plugin API
    const AZsForRegion = [
        'us-east-1a',
        'us-east-1b',
        'us-east-1c',
        'us-east-1d',
        'us-east-1e',
        'us-east-1f',
        'ap-southeast-2a',
        'ap-southeast-2b',
    ];

    // This will be replaced by an API call
    const promiseOptions = (): Promise<Array<OptionType>> =>
        new Promise(resolve => {
            setTimeout(() => {
                resolve(AZsForRegion.map(az => ({ label: az, value: az, key: az })));
            }, 1000);
        });

    const {
        paramProperties: { Description },
    } = param;

    const validate = (value: Array<OptionType>): string => {
        if (value.length !== 2) {
            return 'INCORRECT_NUM_AZ';
        }
        return undefined;
    };

    return (
        <Field validate={validate} {...defaultFieldProps}>
            {({ fieldProps, error }: any): ReactElement => (
                <>
                    <HelperMessage>{Description}</HelperMessage>
                    <AsyncSelect
                        className="az-select"
                        cacheOptions
                        defaultOptions
                        isMulti
                        isSearchable={false}
                        loadOptions={promiseOptions}
                        {...fieldProps}
                    />
                    {error && (
                        <ErrorMessage>
                            {I18n.getText(
                                'atlassian.migration.datacenter.provision.aws.form.wrongNumberAZError'
                            )}
                        </ErrorMessage>
                    )}
                </>
            )}
        </Field>
    );
};

const createNumberInputFromQuickstartParam: FormElementGenerator = (defaultFieldProps, param) => {
    const {
        paramProperties: { Default, ConstraintDescription, Description, MaxValue, MinValue },
    } = param;

    const overrideInputProps: InputProps = {
        type: 'number',
    };

    let overrideFieldProps: InputProps = {
        defaultValue: (Default as number) || '',
    };

    if (MaxValue) {
        overrideFieldProps = {
            ...overrideFieldProps,
            validate: (value: number): string => {
                if (value <= MaxValue) {
                    return undefined;
                }
                return ConstraintDescription;
            },
        };
    }

    if (MinValue) {
        const minValueValidateFunc = (value: number): string => {
            if (value >= MinValue) {
                return undefined;
            }
            return ConstraintDescription;
        };
        let validate = minValueValidateFunc;
        if (MaxValue) {
            const oldValidate = overrideFieldProps.validate as Function;
            validate = (value: number): string => {
                return oldValidate(value) || minValueValidateFunc(value);
            };
        }

        overrideFieldProps = {
            ...overrideFieldProps,
            validate,
        };
    }
    return (
        <Field {...defaultFieldProps} {...overrideFieldProps}>
            {({ fieldProps, error }: any): ReactElement => {
                return (
                    <>
                        <HelperMessage>{Description}</HelperMessage>
                        <TextField width="medium" {...fieldProps} {...overrideInputProps} />
                        {error && <ErrorMessage>{error}</ErrorMessage>}
                    </>
                );
            }}
        </Field>
    );
};

const createStringInputFromQuickstartParam: FormElementGenerator = (defaultFieldProps, param) => {
    const {
        paramProperties: {
            AllowedPattern,
            ConstraintDescription,
            Description,
            Default,
            MaxLength,
            NoEcho,
        },
    } = param;

    let overrideInputProps: Record<string, string | number | boolean | Function> = {
        type: NoEcho ? 'password' : 'text',
    };

    let overrideFieldProps: Record<string, string | number | boolean | Function> = {
        defaultValue: (Default as string) || '',
    };

    if (MaxLength) {
        overrideInputProps = {
            maxLength: MaxLength,
            ...overrideInputProps,
        };
    }

    if (AllowedPattern) {
        overrideInputProps = {
            ...overrideInputProps,
        };
        overrideFieldProps = {
            ...overrideFieldProps,
            validate: (value: string): string => {
                const regex = new RegExp(AllowedPattern);
                const testSuccess = regex.test(value);
                if (testSuccess) {
                    return undefined;
                }
                return (
                    ConstraintDescription ||
                    `${param.paramLabel} ${I18n.getText(
                        'atlassian.migration.datacenter.provision.aws.form.defaultError'
                    )} ${AllowedPattern}`
                );
            },
        };
    }

    return (
        <Field {...defaultFieldProps} {...overrideFieldProps}>
            {({ fieldProps, error }: any): ReactElement => (
                <>
                    <HelperMessage>{Description}</HelperMessage>
                    <TextField width="xlarge" {...fieldProps} {...overrideInputProps} />
                    {error && <ErrorMessage>{error}</ErrorMessage>}
                </>
            )}
        </Field>
    );
};

const createInputFromQuickstartParam: FormElementGenerator = (defaultFieldProps, param) => {
    const {
        paramKey,
        paramProperties: { Type },
    } = param;
    if (Type === 'Number') {
        return createNumberInputFromQuickstartParam(defaultFieldProps, param);
    }
    if (Type === 'String') {
        return createStringInputFromQuickstartParam(defaultFieldProps, param);
    }

    return <div key={paramKey}>UNRECOGNISED PARAM TYPE</div>;
};

const createSelectFromQuickstartParam: FormElementGenerator = (defaultFieldProps, param) => {
    const { paramProperties } = param;
    const { AllowedValues, Default, Description } = paramProperties;
    if (AllowedValues.length === 2 && typeof AllowedValues[0] === 'boolean') {
        return (
            <Field {...defaultFieldProps}>
                {({ fieldProps }: any): ReactElement => (
                    <>
                        <HelperMessage>{Description}</HelperMessage>
                        <Toggle
                            {...fieldProps}
                            size="large"
                            isDefaultChecked={Default as boolean}
                        />
                    </>
                )}
            </Field>
        );
    }

    const options = AllowedValues.map(val => ({ label: val as string, value: val, key: val }));
    const defaultOption = options.find(option => option.value === Default);

    const overrideFieldProps = {
        options,
    };

    return (
        <Field {...defaultFieldProps} defaultValue={defaultOption}>
            {({ fieldProps }: any): ReactElement => (
                <>
                    <HelperMessage>{Description}</HelperMessage>
                    <Select {...fieldProps} {...overrideFieldProps} />
                </>
            )}
        </Field>
    );
};

const quickstartParamToAtlaskitFormElement: FormElementGenerator = (defaultFieldProps, param) => {
    const { paramProperties } = param;
    if (paramProperties.AllowedValues) {
        return createSelectFromQuickstartParam(defaultFieldProps, param);
    }
    if (paramProperties.Type === 'List<AWS::EC2::AvailabilityZone::Name>') {
        return createAZSelection(defaultFieldProps, param);
    }
    return createInputFromQuickstartParam(defaultFieldProps, param);
};

export const createQuickstartFormField = (param: QuickstartParameter): ReactElement => {
    const { paramKey, paramLabel } = param;

    const defaultFieldProps = {
        key: paramKey,
        label: paramLabel,
        name: paramKey,
        defaultValue: '',
    };

    return quickstartParamToAtlaskitFormElement(defaultFieldProps, param);
};
