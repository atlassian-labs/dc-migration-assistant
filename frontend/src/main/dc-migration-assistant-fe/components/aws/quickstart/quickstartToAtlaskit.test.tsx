/*
 * Copyright 2020 Atlassian
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { render } from '@testing-library/react';

// eslint-disable-next-line import/extensions
import { QuickstartParameter } from './QuickStartTypes';
import { createQuickstartFormField } from './quickstartToAtlaskit';

const TEST_PARAM_KEY = 'TestParam';
const TEST_PARAM_LABEL = 'RANDOM_TEST_PARAMETER';
const TEST_PARAM_DESCRIPTION = 'a test parameter';
const TEST_STRING_DEFAULT = 'test default';
const TEST_NUMBER_DEFAULT = 12;

const STRING_INPUT: QuickstartParameter = {
    paramKey: TEST_PARAM_KEY,
    paramLabel: TEST_PARAM_LABEL,
    paramProperties: {
        Type: 'String',
        Default: TEST_STRING_DEFAULT,
        Description: TEST_PARAM_DESCRIPTION,
    },
};

const NO_ECHO_INPUT: QuickstartParameter = {
    ...STRING_INPUT,
    paramProperties: {
        ...STRING_INPUT.paramProperties,
        NoEcho: true,
    },
};

const NUMBER_INPUT: QuickstartParameter = {
    paramKey: TEST_PARAM_KEY,
    paramLabel: TEST_PARAM_LABEL,
    paramProperties: {
        Type: 'Number',
        Default: TEST_NUMBER_DEFAULT,
        Description: TEST_PARAM_DESCRIPTION,
    },
};

const TOGGLE_INPUT: QuickstartParameter = {
    paramKey: TEST_PARAM_KEY,
    paramLabel: TEST_PARAM_LABEL,
    paramProperties: {
        Type: 'String',
        Default: true,
        Description: TEST_PARAM_DESCRIPTION,
        AllowedValues: [true, false],
    },
};

const SELECT_INPUT: QuickstartParameter = {
    paramKey: TEST_PARAM_KEY,
    paramLabel: TEST_PARAM_LABEL,
    paramProperties: {
        Type: 'String',
        Default: 'Option 1',
        Description: TEST_PARAM_DESCRIPTION,
        AllowedValues: ['Option 1', 'Option 2'],
    },
};

const AZ_INPUT: QuickstartParameter = {
    paramKey: TEST_PARAM_KEY,
    paramLabel: TEST_PARAM_LABEL,
    paramProperties: {
        Type: 'List<AWS::EC2::AvailabilityZone::Name>',
        Description: TEST_PARAM_DESCRIPTION,
    },
};

const mockFetch = jest.fn();
mockFetch.mockReturnValue(Promise.resolve({ json: ['us-east-1a', 'us-east-1b'] }));
window.fetch = mockFetch;

describe('Quick Start to Atlaskit Input Converter', () => {
    it('Should create a label and description for any parameter', () => {
        const { getByText } = render(createQuickstartFormField(STRING_INPUT));

        const label = getByText(TEST_PARAM_LABEL);
        expect(label).toBeTruthy();

        const description = getByText(TEST_PARAM_DESCRIPTION);
        expect(description).toBeTruthy();
    });

    it('Should create a number input', () => {
        const { getByLabelText } = render(createQuickstartFormField(NUMBER_INPUT));

        expect(getByLabelText(TEST_PARAM_LABEL)).toMatchSnapshot();
    });

    it('Should render String inputs', () => {
        const { getByLabelText } = render(createQuickstartFormField(STRING_INPUT));

        const stringInput = getByLabelText(TEST_PARAM_LABEL);
        expect(stringInput).toMatchSnapshot();
    });

    it('Should create a toggle input when there are two allowed values that are boolean', () => {
        const { getByRole } = render(createQuickstartFormField(TOGGLE_INPUT));

        // Atlaskit toggle does not use the label attributes properly so we must look for a checkbox
        expect(getByRole('checkbox')).toMatchSnapshot();
    });

    it('Should create a select input when there are multiple allowed values that are not boolean', () => {
        const { getAllByLabelText } = render(createQuickstartFormField(SELECT_INPUT));

        const selectComponents = getAllByLabelText(TEST_PARAM_LABEL);
        const select = selectComponents.find(component => component.id === 'react-select-2-input');

        expect(select).toMatchSnapshot();
    });

    it('Should create multi select input for AZ parameter', () => {
        const { container } = render(createQuickstartFormField(AZ_INPUT));

        expect(container.querySelector('az-select')).toMatchSnapshot();
    });

    it('Should create a password input for NoEcho parameters', () => {
        const { getByLabelText } = render(createQuickstartFormField(NO_ECHO_INPUT));

        expect(getByLabelText(TEST_PARAM_LABEL)).toMatchSnapshot();
    });

    it('Should populate parameters with their default values', () => {
        const { getByLabelText } = render(createQuickstartFormField(NUMBER_INPUT));

        expect(getByLabelText(TEST_PARAM_LABEL).getAttribute('value')).toEqual(
            `${TEST_NUMBER_DEFAULT}`
        );
    });
});
