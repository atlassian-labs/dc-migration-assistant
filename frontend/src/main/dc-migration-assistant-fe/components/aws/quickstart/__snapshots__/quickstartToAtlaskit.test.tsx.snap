// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quick Start to Atlaskit Input Converter Should create a number input 1`] = `
<input
  aria-invalid="false"
  aria-labelledby="TestParam-uid2-label TestParam-uid2-helper TestParam-uid2-valid TestParam-uid2-error"
  class="css-tonkbf-Input"
  id="TestParam-uid2"
  name="TestParam"
  type="number"
  value="12"
/>
`;

exports[`Quick Start to Atlaskit Input Converter Should create a password input for NoEcho parameters 1`] = `
<input
  aria-invalid="false"
  aria-labelledby="TestParam-uid8-label TestParam-uid8-helper TestParam-uid8-valid TestParam-uid8-error"
  class="css-tonkbf-Input"
  id="TestParam-uid8"
  name="TestParam"
  type="password"
  value="test default"
/>
`;

exports[`Quick Start to Atlaskit Input Converter Should create a select input when there are multiple allowed values that are not boolean 1`] = `
<input
  aria-autocomplete="list"
  aria-labelledby="TestParam-uid6-label TestParam-uid6-helper TestParam-uid6-valid TestParam-uid6-error"
  autocapitalize="none"
  autocomplete="off"
  autocorrect="off"
  id="react-select-2-input"
  spellcheck="false"
  style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
  tabindex="0"
  type="text"
  value=""
/>
`;

exports[`Quick Start to Atlaskit Input Converter Should create a toggle input when there are two allowed values that are boolean 1`] = `
<input
  checked=""
  class="sc-iwsKbI bCvnPm"
  id="uid5"
  name="TestParam"
  type="checkbox"
  value=""
/>
`;

exports[`Quick Start to Atlaskit Input Converter Should create multi select input for AZ parameter 1`] = `null`;

exports[`Quick Start to Atlaskit Input Converter Should render String inputs 1`] = `
<input
  aria-invalid="false"
  aria-labelledby="TestParam-uid3-label TestParam-uid3-helper TestParam-uid3-valid TestParam-uid3-error"
  class="css-tonkbf-Input"
  id="TestParam-uid3"
  name="TestParam"
  type="text"
  value="test default"
/>
`;
